File: src/main/resources/templates/swift/xcode/solution/${appName}/Sort/MergeSort.swift
//
//  MergeSort.swift
//  ${appName}
//

import Foundation

public class MergeSort: SortStrategy {
    /**
     Sorts dates with MergeSort.

     - Parameter input: the List of Dates to be sorted
    */
    
    /*
     Task 1.2:
     Implement the Merge Sort algorithm
    */
    public func performSort(_ input: [Date]) -> [Date] {
        return mergesort(input, 0, input.count - 1)
    }

    /// Recursive merge sort method
    private func mergesort(_ input: [Date], _ low: Int, _ high: Int) -> [Date] {
        var sortedInput = input
        if (high - low) < 1 {
            /// break recursion and return last input
            return input
        }
        let mid: Int = (low + high) / 2
        sortedInput = mergesort(sortedInput, low, mid)
        sortedInput = mergesort(sortedInput, mid + 1, high)
        sortedInput = merge(sortedInput, low, mid, high)
        return sortedInput
    }

    /// Merge method
    private func merge(_ input: [Date], _ low: Int, _ middle: Int, _ high: Int) -> [Date] {
        var sortedInput = input
        var temp: [Date] = [Date]()
        var leftIndex: Int = low
        var rightIndex: Int = middle + 1
        while (leftIndex <= middle) && (rightIndex <= high) {
            if sortedInput[leftIndex] <= sortedInput[rightIndex] {
                temp.append(sortedInput[leftIndex])
                leftIndex += 1
            } else {
                temp.append(sortedInput[rightIndex])
                rightIndex += 1
            }
        }

        if (leftIndex <= middle) && (rightIndex > high) {
            while leftIndex <= middle {
                temp.append(sortedInput[leftIndex])
                leftIndex += 1
            }
        } else {
            while rightIndex <= high {
                temp.append(sortedInput[rightIndex])
                rightIndex += 1
            }
        }
        for wholeIndex in 0 ..< temp.count {
            sortedInput[wholeIndex + low] = temp[wholeIndex]
        }
        return sortedInput
    }
}

File: src/main/resources/templates/c_sharp/solution/MergeSort.cs
namespace assignment;

public class MergeSort : ISortStrategy
{

    /// <summary>
    /// Wrapper method for the real MergeSort algorithm.
    /// </summary>
    /// <param name="input">the List of Dates to be sorted</param>
    public void PerformSort(List<DateTime> input)
    {
        Mergesort(input, 0, input.Count - 1);
    }

    /// <summary>
    /// Recursive merge sort method
    /// </summary>
    private void Mergesort(List<DateTime> input, int low, int high)
    {
        if (high - low < 1)
        {
            return;
        }

        int mid = (low + high) / 2;

        Mergesort(input, low, mid);
        Mergesort(input, mid + 1, high);
        Merge(input, low, mid, high);
    }

    /// <summary>
    /// Merge method
    /// </summary>
    private void Merge(List<DateTime> input, int low, int middle, int high)
    {
        DateTime[] temp = new DateTime[high - low + 1];

        int leftIndex = low;
        int rightIndex = middle + 1;
        int wholeIndex = 0;

        while (leftIndex <= middle && rightIndex <= high)
        {
            if (input[leftIndex] <= input[rightIndex])
            {
                temp[wholeIndex] = input[leftIndex++];
            }
            else
            {
                temp[wholeIndex] = input[rightIndex++];
            }
            wholeIndex++;
        }

        while (leftIndex <= middle)
        {
            temp[wholeIndex++] = input[leftIndex++];
        }
        while (rightIndex <= high)
        {
            temp[wholeIndex++] = input[rightIndex++];
        }

        for (wholeIndex = 0; wholeIndex < temp.Length; wholeIndex++)
        {
            input[wholeIndex + low] = temp[wholeIndex];
        }
    }
}
